exercises/rollback-with-saga/solution/workflow.go

```diff
@@ -0,0 +1,94 @@
+package pizza
+
+import (
+	"errors"
+	"time"
+
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+	"go.uber.org/multierr"
+)
+
+func PizzaWorkflow(ctx workflow.Context, order PizzaOrder) (OrderConfirmation, error) {
+	retrypolicy := &temporal.RetryPolicy{
+		MaximumInterval: time.Second * 10,
+		MaximumAttempts: 3,
+	}
+
+	options := workflow.ActivityOptions{
+		StartToCloseTimeout: time.Second * 5,
+		HeartbeatTimeout:    10 * time.Second,
+		RetryPolicy:         retrypolicy,
+	}
+
+	ctx = workflow.WithActivityOptions(ctx, options)
+
+	logger := workflow.GetLogger(ctx)
+
+	var totalPrice int
+	for _, pizza := range order.Items {
+		totalPrice += pizza.Price
+	}
+
+	var distance Distance
+	err := workflow.ExecuteActivity(ctx, GetDistance, order.Address).Get(ctx, &distance)
+	if err != nil {
+		logger.Error("Unable to get distance", "Error", err)
+		return OrderConfirmation{}, err
+	}
+
+	if order.IsDelivery && distance.Kilometers > 12 {
+		return OrderConfirmation{}, errors.New("Out of Service Area")
+	}
+
+	err = workflow.ExecuteActivity(ctx, UpdateInventory, order.Items).Get(ctx, nil)
+	if err != nil {
+		return OrderConfirmation{}, err
+	}
+
+	defer func() {
```

Comment:
Okay not too sure about this one - but shouldn't we remove defer? From  Go docs, "A defer statement defers the execution of a function until the surrounding function return",  so my concern is that the deferred function will not execute the compensation logic immediateley when the error gets caught. Instead, it might waits until the function returns.

Reply:
Right, that's the idea -- it tracks all the errors up to that point and then rolls them all back when the Workflow returns. That's how the sample does it, anyhow: https://github.com/temporalio/samples-go/blob/main/saga/workflow.go

Reply:
Ah okay - I thought we wanted to immediately execute when the error gets caught, thanks for letting me know :)